mesh/pncmesh.cpp

```diff
@@ -577,41 +577,81 @@ void ParNCMesh::CalcFaceOrientations()
    face_orient.SetSize(NFaces);
    face_orient = 0;
 
-   for (auto face = faces.begin(); face != faces.end(); ++face)
+   for (auto face : faces)
    {
-      if (face->elem[0] >= 0 && face->elem[1] >= 0 && face->index < NFaces)
+      if (face.elem[0] >= 0 && face.elem[1] >= 0 && face.index < NFaces)
       {
-         Element *e1 = &elements[face->elem[0]];
-         Element *e2 = &elements[face->elem[1]];
+         Element *e1 = &elements[face.elem[0]];
+         Element *e2 = &elements[face.elem[1]];
 
          if (e1->rank == e2->rank) { continue; }
          if (e1->rank > e2->rank) { std::swap(e1, e2); }
 
-         face_orient[face->index] = get_face_orientation(*face, *e1, *e2);
+         face_orient[face.index] = get_face_orientation(face, *e1, *e2);
       }
    }
 }
 
 void ParNCMesh::GetBoundaryClosure(const Array<int> &bdr_attr_is_ess,
                                    Array<int> &bdr_vertices,
-                                   Array<int> &bdr_edges)
+                                   Array<int> &bdr_edges, Array<int> &bdr_faces)
 {
-   NCMesh::GetBoundaryClosure(bdr_attr_is_ess, bdr_vertices, bdr_edges);
+   NCMesh::GetBoundaryClosure(bdr_attr_is_ess, bdr_vertices, bdr_edges, bdr_faces);
 
-   int i, j;
-   // filter out ghost vertices
-   for (i = j = 0; i < bdr_vertices.Size(); i++)
+   if (Dim == 3)
    {
-      if (bdr_vertices[i] < NVertices) { bdr_vertices[j++] = bdr_vertices[i]; }
+      // Mark masters of shared slave boundary faces as essential boundary faces. Some
+      // master faces may only have slave children.
+      for (const auto &mf : shared_faces.masters)
+      {
+         if (elements[mf.element].rank != MyRank) { continue; }
+         for (int j = mf.slaves_begin; j < mf.slaves_end; j++)
+         {
+            const auto &sf = GetFaceList().slaves[j];
+            if (sf.index < 0)
+            {
+               // Edge-face constraint. Skip this edge.
+               continue;
+            }
+            Face *face = GetFace(elements[sf.element], sf.local);
+            if (face && face->Boundary() && bdr_attr_is_ess[face->attribute - 1])
```

Comment:
Out of curiosity, and maybe to prompt more detailed documentation, in what case would `face` be null? Is it the case of this slave being a master in the hierarchy, so it is not in the mesh?

Reply:
I'm not aware of a case where it would be. As the local face numbers on the slaves should always be valid. I was using the `GetFace` api that returned a pointer by default however, and whenever accessing a pointer I put the check out of cleanliness. In this case however I can force the survival of the temporary and bind to `const&` and not worry given the extra information I have on the validity of the arguments.